router/router_test.go

```diff
@@ -108,3 +104,129 @@ func TestRingpopRouterMakeRemoteClientCached(t *testing.T) {
 	router.GetClient("hello")
 	cf.AssertNumberOfCalls(t, "MakeRemoteClient", 1)
 }
+
+func TestRingpopRouterRemoveExistingClient(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.removeClient(hostport)
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveNonExistingClient(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	assert.Equal(t, nil, router.clientCache[hostport])
+	router.removeClient(hostport)
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveClientOnSwimFaultyEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.HandleEvent(swim.MemberlistChangesReceivedEvent{
+		Changes: []swim.Change{
+			swim.Change{
+				Address: hostport,
+				Status:  swim.Faulty,
+			},
+		},
+	})
+
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveClientOnSwimLeaveEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.HandleEvent(swim.MemberlistChangesReceivedEvent{
+		Changes: []swim.Change{
+			swim.Change{
+				Address: hostport,
+				Status:  swim.Leave,
+			},
+		},
+	})
+
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterNotRemoveClientOnSwimSuspectEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
```

Comment:
We don't need any extra methods to set a specific key on a specific value in the cache. For this we can callout to `GetClient`. But that makes this and the other tests rely on the correctness of `GetClient` which is tested elsewhere. I used to do that in the past in similar situations, but it is hard to read from the test result that the only function failing is `GetClient` opposed to the invalidation of the cache. Therefor I now fall back to manually setting the initial state via access to the private members.

If `clientCache` gets refactored it will be clear from the compile error on this line.